epochStart/bootstrap/shardStorageHandler.go

```diff
@@ -177,31 +177,75 @@ func getEpochStartShardData(metaBlock data.MetaHeaderHandler, shardId uint32) (d
 	return &block.EpochStartShardData{}, epochStart.ErrEpochStartDataForShardNotFound
 }
 
-func (ssh *shardStorageHandler) getCrossProcessedMbsDestMeByHeader(
-	shardHeader data.ShardHeaderHandler,
-) map[uint32][]data.MiniBlockHeaderHandler {
-	crossMbsProcessed := make(map[uint32][]data.MiniBlockHeaderHandler)
+func (ssh *shardStorageHandler) getCrossProcessedMiniBlockHeadersDestMe(shardHeader data.ShardHeaderHandler) map[string]data.MiniBlockHeaderHandler {
+	crossMbsProcessed := make(map[string]data.MiniBlockHeaderHandler)
 	processedMiniBlockHeaders := shardHeader.GetMiniBlockHeaderHandlers()
 	ownShardID := shardHeader.GetShardID()
 
-	for i, mbHeader := range processedMiniBlockHeaders {
+	for index, mbHeader := range processedMiniBlockHeaders {
 		if mbHeader.GetReceiverShardID() != ownShardID {
 			continue
 		}
 		if mbHeader.GetSenderShardID() == ownShardID {
 			continue
 		}
 
-		mbs, ok := crossMbsProcessed[mbHeader.GetSenderShardID()]
+		crossMbsProcessed[string(mbHeader.GetHash())] = processedMiniBlockHeaders[index]
+	}
+
+	return crossMbsProcessed
+}
+
+func getProcessedMiniBlocksForFinishedMeta(
+	referencedMetaBlockHashes [][]byte,
+	headers map[string]data.HeaderHandler,
+	selfShardID uint32,
+) ([]bootstrapStorage.MiniBlocksInMeta, error) {
+
+	processedMiniBlocks := make([]bootstrapStorage.MiniBlocksInMeta, 0)
+
+	for i := 0; i < len(referencedMetaBlockHashes)-1; i++ {
+		header, ok := headers[string(referencedMetaBlockHashes[i])]
 		if !ok {
-			mbs = make([]data.MiniBlockHeaderHandler, 0)
+			return nil, fmt.Errorf("%w in getProcessedMiniBlocksForFinishedMeta: hash: %s",
+				epochStart.ErrMissingHeader,
+				hex.EncodeToString(referencedMetaBlockHashes[i]))
+		}
+
+		neededMeta, ok := header.(*block.MetaBlock)
+		if !ok {
+			return nil, epochStart.ErrWrongTypeAssertion
+		}
+		if check.IfNil(neededMeta) {
+			return nil, epochStart.ErrNilMetaBlock
 		}
 
-		mbs = append(mbs, processedMiniBlockHeaders[i])
-		crossMbsProcessed[mbHeader.GetSenderShardID()] = mbs
+		log.Debug("getProcessedMiniBlocksForFinishedMeta", "meta block hash", referencedMetaBlockHashes[i])
+
+		miniBlockHashes := make([][]byte, 0)
+		isFullyProcessed := make([]bool, 0)
+		indexOfLastTxProcessed := make([]int32, 0)
+
+		miniBlockHeadersDestMe := getMiniBlockHeadersForDest(neededMeta, selfShardID)
+		for mbHash, mbHeader := range miniBlockHeadersDestMe {
+			log.Debug("getProcessedMiniBlocksForFinishedMeta", "mb hash", mbHash)
+
+			miniBlockHashes = append(miniBlockHashes, []byte(mbHash))
+			isFullyProcessed = append(isFullyProcessed, mbHeader.IsFinal())
```

Comment:
This is done when the mini block header is created by proposer and then validated by the consensus. In case of inconsistency the block will be dropped at that point

Reply:
Added another consistency check in the validator side in ProcessBlock -> checkHeaderBodyCorrelation method